src/v2.rs

```diff
@@ -0,0 +1,280 @@
+    result.extend_from_slice(ESZIP_V2_MAGIC);
+    result.extend_from_slice(&header_len.to_be_bytes());
+    result.extend_from_slice(&header);
+    result.extend_from_slice(&self.data);
+    result
+  }
+
+  pub fn get_module<'a>(&'a self, specifier: &str) -> Option<Module> {
+    let mut specifier = specifier;
+    let mut visited = HashSet::new();
+    loop {
+      visited.insert(specifier);
+      let module = self.header.modules.get(specifier)?;
+      match module {
+        EsZipV2Module::Redirect { redirect } => {
+          specifier = redirect.as_str();
+          if visited.contains(specifier) {
+            return None;
+          }
+        }
+        EsZipV2Module::Source {
+          offset,
+          size,
+          r#type,
+        } => {
+          let offset = *offset as usize;
+          let size = *size as usize;
+          let source = &self.data[offset..offset + size];
+          let module = Module {
+            specifier: specifier.to_owned(),
+            source: Cow::Borrowed(source),
+            r#type: r#type.clone(),
+          };
+          return Some(module);
+        }
+      }
+    }
+  }
+
+  pub fn from_graph(graph: ModuleGraph) -> Result<Self, anyhow::Error> {
+    let mut eszip = EsZipV2::default();
+
+    for (specifier, redirect) in &graph.redirects {
+      eszip.header.modules.insert(
+        specifier.to_string(),
+        EsZipV2Module::Redirect {
+          redirect: redirect.to_string(),
+        },
+      );
+    }
+
+    for module in graph.synthetic_modules() {
+      let specifier = module.specifier.to_string();
+      if module.media_type == MediaType::Json {
+        let source = module.maybe_source.as_ref().unwrap();
+        let data = source.as_bytes();
+        let offset = eszip.data.len() as u64;
+        eszip.data.extend_from_slice(data);
+        eszip.header.modules.insert(
+          specifier,
+          EsZipV2Module::Source {
+            size: data.len() as u64,
+            offset,
+            r#type: ModuleType::JSON,
+          },
+        );
+      }
+    }
+
+    for module in graph.modules() {
+      let emit_options = EmitOptions::default();
+      let (emit, _) = ast::transpile(&module.parsed_source, &emit_options)?;
+      let specifier = module.specifier.to_string();
+      let data = emit.as_bytes();
+      let offset = eszip.data.len() as u64;
+      eszip.data.extend_from_slice(data);
+      eszip.header.modules.insert(
+        specifier,
+        EsZipV2Module::Source {
+          size: data.len() as u64,
+          offset,
+          r#type: ModuleType::JS,
+        },
+      );
+    }
+
+    Ok(eszip)
+  }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+struct EsZipV2Header {
+  modules: HashMap<String, EsZipV2Module>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+enum EsZipV2Module {
+  Redirect {
+    redirect: String,
+  },
+  Source {
+    size: u64,
+    offset: u64,
+    r#type: ModuleType,
```

Comment:
```suggestion
    #[serde(rename = "type")]
    type_: ModuleType,
```
Maybe? Many crates have been using this pattern.

Reply:
`kind` though tends to better all around if you can get away with it.